### PR TITLE
ConvLayer: add pad_seq_len_to_power

### PR DESCRIPTION
**Pull request by vieting**
We encountered a problem with linearly growing memory here https://github.com/rwth-i6/returnn/issues/1450. It seems to be cause by a problem in tensorflow https://github.com/tensorflow/tensorflow/issues/62441. Many different input shapes for the convolution lead to a large CPU memory consumption. As a workaround, we can introduce padding in the `ConvLayer` to reduce the number of different sequence lengths/input shapes. This reduces the CPU memory consumption. After the convolution, the padding can be removed again.

### CODE REVIEW

File: returnn/tf/layers/basic.py [hunk 1]
```diff
@@ -6361,6 +6378,11 @@ def __init__(
             )
         if num_batch_dims > 1:
             y = tf.reshape(y, tf.concat([extended_batch_shape, tf.shape(y)[1:]], axis=0))
+
+        if pad_seq_len_to_power is not None:
+            for ax in self.output.get_dynamic_axes():
+                y = tf.gather(y, tf.range(tf.reduce_max(self.output.get_dynamic_size(ax))), axis=ax)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:6384]: Use `tf.slice`.

File: returnn/tf/layers/basic.py [hunk 2]
```diff
@@ -6301,6 +6304,20 @@ def __init__(
         else:
             x = input_data.placeholder
 
+        if pad_seq_len_to_power is not None:
+            padding_for_power = []
+            for ax in range(input_data.batch_ndim):
+                if input_data.is_axis_dynamic(ax):
```
Comment (albertz) [inline returnn/tf/layers/basic.py:6310]: In your previous test, I think you only padded the spatial dims, not the batch dim. Is this intended that you also pad the batch dim here?

Comment (albertz) [reply returnn/tf/layers/basic.py:6310]: Oh you are right. `is_axis_dynamic` is a bit confusing.

Comment (albertz) [reply returnn/tf/layers/basic.py:6310]: In your test, it was already constant mem consumption (after a while of running) when only doing this?

Comment (vieting) [reply returnn/tf/layers/basic.py:6310]: I started a training for my real config with this option in the first two layers. However, there are more `ConvLayers`. I could add it to all or I could also create a config for testing e.g. with an adapted `Task12AXDataset`.

File: returnn/tf/layers/basic.py [hunk 3]
```diff
@@ -6158,6 +6159,8 @@ def __init__(
         :param dict[str,str]|None filter_perm: transposes the filter (input filter as layer)
         :param LayerBase|None bias: if given, will not create an own parameter, but use this as the bias
         :param bool use_time_mask:
+        :param Optional[float] pad_seq_len_to_power: pad seq len to power of given number to reduce number of different
+            seq lens. See RETURNN #1450 and https://github.com/tensorflow/tensorflow/issues/62441.
```
Comment (albertz) [inline returnn/tf/layers/basic.py:6163]:
```suggestion
        :param Optional[float] pad_seq_len_to_power: pad seq len to power of given number
            to reduce number of different seq lens.
            See https://github.com/rwth-i6/returnn/issues/1450 and https://github.com/tensorflow/tensorflow/issues/62441.
```

**Comment by vieting** [comment]
Btw, if I see it correctly, the pycharm-inspect test only fails for torch related files which are not affected by this PR.

File: returnn/tf/layers/basic.py (hunk 2 above)
Comment (albertz) [reply returnn/tf/layers/basic.py:6310]: Is it needed for all? This was not clear yet, or was it? Or actually I thought in your other network where you only have the other layers, this problem does not occur?

**Comment by albertz** [comment]
> Btw, if I see it correctly, the pycharm-inspect test only fails for torch related files which are not affected by this PR.

It's already fixed, maybe rebase or merge.

File: returnn/tf/layers/basic.py (hunk 2 above)
Comment (vieting) [reply returnn/tf/layers/basic.py:6311]: I think the other layers also contribute to the memory increase. However, since they operate on subsampled sequences, the number of different input shape is not so large, so the increase is not as significant as on the waveform.

File: returnn/tf/layers/basic.py [hunk 4]
```diff
@@ -6158,6 +6159,9 @@ def __init__(
         :param dict[str,str]|None filter_perm: transposes the filter (input filter as layer)
         :param LayerBase|None bias: if given, will not create an own parameter, but use this as the bias
         :param bool use_time_mask:
+        :param Optional[float] pad_seq_len_to_power: pad seq len to power of given number to reduce number of different
+            seq lens. See https://github.com/rwth-i6/returnn/issues/1450 and
+            https://github.com/tensorflow/tensorflow/issues/62441.
```
Comment (albertz) [inline returnn/tf/layers/basic.py:6164]: "seq lens" should not be together in the same line with the next sentence, if the next sentence is partially split over again more lines.
```suggestion
        :param Optional[float] pad_seq_len_to_power: pad seq len to power of given number
            to reduce number of different seq lens.
            See https://github.com/rwth-i6/returnn/issues/1450 and
            https://github.com/tensorflow/tensorflow/issues/62441.
```

File: returnn/tf/layers/basic.py (hunk 2 above)
Comment (albertz) [reply returnn/tf/layers/basic.py:6311]: Ah ok. So if this is a problem for all ConvLayers, instead of having this a per-layer option, this could also be a global option in the config.

Comment (vieting) [reply returnn/tf/layers/basic.py:6312]: How would you do that? Can I access the global config from within a layer?

Comment (vieting) [reply returnn/tf/layers/basic.py:6312]: `self.network._config` might work but is there a nice way to access it?

Comment (vieting) [reply returnn/tf/layers/basic.py:6312]: Something like `pad_seq_len_to_power = pad_seq_len_to_power or self.network.get_config().float("conv_pad_seq_len_to_power", None)`?

Comment (albertz) [reply returnn/tf/layers/basic.py:6315]: Yes `self.network.get_config()`. But I think I would also remove the layer option completely then (unless you think it makes sense to also do this per layer).

Comment (vieting) [reply returnn/tf/layers/basic.py:6315]: It depends a bit on how much this affects the performance I guess. Also, you might want to specify lower values here for initial layers (where sequence lengths are tens of thousands) and higher values for layers after subsampling.

Comment (albertz) [reply returnn/tf/layers/basic.py:6315]: It's logarithmic, so shouldn't it be the same?
But yes ok, if you think it's useful for you, then add it.

Comment (vieting) [reply returnn/tf/layers/basic.py:6315]: It only adds minimally to the code complexity, so I think I'd keep it.

**Comment by vieting** [comment]
I compared padding the time dim for the first two layers on the waveform or for all `ConvLayer`s and additionally also padding the batch dim.

image

There is clearly an advantage when padding all `ConvLayer`s, however, padding the batch dim hardly makes any difference. In terms of training speed, padding all layers is ~6.3% slower than just the first two. Padding time + batch is ~10.5% slower than just padding time. The small reduction of memory does not seem to justify the slowdown in case of batch padding imho. If you still want to add it, I'd at least make it configurable.

We can also see that the linear memory growth is not yet fixed. There might be some other layer which leaks memory. However, the padding introduced in this PR largely helps reducing the leak. I'd therefore vote to merge it so it's already available and we might discuss other measures in a separate issue/PR.

**Review by albertz — APPROVED** [review]
